### ISSUE TITLE
Migration Tool treats local file as an external package that hasn't migrated

### ISSUE THREAD

**Issue by escamoteur** [dart-lang/sdk#45780]
Please see this screenshot:
![image](https://user-images.githubusercontent.com/16635729/115529389-2899ea80-a293-11eb-83dd-34327ef85e26.png)

```yaml
name: listenable_collections
description: A collection of Dart collections that behave like ValueNotifier if their data changes.
version: 0.0.5
author: Thomas Burkhart
homepage: https://github.com/escamoteur/listenable_collections

environment:
  sdk: ">=2.7.0 <3.0.0"
  flutter: ">=1.17.0 <2.0.0"

dependencies:
  collection: ^1.15.0
  flutter:
    sdk: flutter
```
this is the repo I try to migrate https://github.com/escamoteur/listenable_collections

**Comment by stereotype441** [comment]
Hi @escamoteur, thanks for the bug report!  It looks like you aren't the first person to run into this problem - a similar issue was also reported in #45478, which I unfortunately wasn't able to reproduce.  Fortunately, I was able to reproduce your bug (checking out https://github.com/escamoteur/listenable_collections/commit/010b5ba881a9a7ea99bf5dbf072b6a3a25e1621c), so I will work on a fix.

I see from https://github.com/escamoteur/listenable_collections/commit/0f28b6b945764daf5667aabe63557a9dc2beba21 that you were able to perform migration in spite of this bug.  I'm glad to see that!

In case it's useful to anyone else, I was able to work around the bug and get the migration tool to proceed by changing the `import` statement in `test/list_notifier_test.dart` from this:
```dart
import '../lib/src/list_notifier.dart';
```
to this:
```dart
import 'package:listenable_collections/src/list_notifier.dart';
```

It looks like the migration tool is getting confused by the presence of the relative import from `test` to `lib`.  Most Dart code I work with uses `package:` imports when going from `test` to `lib`, but I'm not sure whether that's just a matter of personal preference or if there's a good reason to prefer one of the other.  I'm talking with some folks on the language team to try to get clarification on that, and I'll update this bug when I have more information.

**Comment by stereotype441** [comment]
Found the problem: the migration tool's logic for detecting unmigrated dependencies uses URIs to exclude a package's own files from the check.  But importing a `lib` file from `test` causes that file to be reached via a `file:` URI, whereas the rest of the migration tool assumes that files in `lib` will be reached via a `package:` URI.  The URIs don't match so the migration tool doesn't exclude the dependency.

I have a fix in the works.

**Comment by dcoleyoung** [comment]
@stereotype441 was there a fix included in the sdk for this?  I seem to be having this issue now.  I'm about three years behind and the migration tool was left out of the latest upgrades.  I'm here

Flutter 3.7.12 • channel unknown • unknown source
Framework • revision 4d9e56e694 (5 months ago) • 2023-04-17 21:47:46 -0400
Engine • revision 1a65d409c7
Tools • Dart 2.19.6 • DevTools 2.20.1

**Comment by stereotype441** [comment]
> @stereotype441 was there a fix included in the sdk for this? I seem to be having this issue now. I'm about three years behind and the migration tool was left out of the latest upgrades. I'm here
> 
> Flutter 3.7.12 • channel unknown • unknown source Framework • revision 4d9e56e694 (5 months ago) • 2023-04-17 21:47:46 -0400 Engine • revision 1a65d409c7 Tools • Dart 2.19.6 • DevTools 2.20.1

Yes, the fix for this was included in Dart 2.14.

**Comment by dcoleyoung** [comment]
> @dcoleyoung @stereotype441 i'm hitting this as well with Dart 2.19.6, could it be possible it's not in there?
> 
> any file that is included with a 'package:<path_to_subdir_in project>/file' is getting flagged by tool...
> 
> thanks

I actually got the conversion tool to work on my app.  What I did was download and use FVM trying it from earlier possible sdk to see what would run.  My app was created with SDK 1.22.0 so it was too far behind for anything modern.

The migration tool ran on sdk 2.0.0 on my app with minimal evaluation errors (~10).  When I corrected those by hand the migration tool ran and I was able to apply the changes.  Then I reboosted the sdk to higher versions.
